### ISSUE TITLE
Composing multiple configuration files

### ISSUE THREAD

**Issue by cihati** [python/mypy#9620]
**Feature**

Ability to specify a base configuration file and another one to override/extend behavior.

**Pitch**

We have a monorepo with Gradle. There are a few different Python projects and each of them define their linters either fully duplicated or only with slight differences. I'd like to standardize this. For mypy, I'd like to have a base configuration file where I define common settings, and then make it possible to have projects overriding/extending behavior.

For instance, one project uses Django, and it has django-stubs plugin for mypy. If I put this into the base config, then mypy complains on the non-Django projects as the plugin becomes unusable. I do know that I can override some configuration behavior through command line arguments (_"Settings override mypy’s built-in defaults and command line flags can override settings."_) but I'm not sure if it's possible for plugin settings anyways.

**Comment by The-Compiler** [comment]
We have a similar situation in [qutebrowser](https://qutebrowser.org/) (cc @toofar), where we want to support different major versions of a library via something like:

```python3
if machinery.USE_PYQT5:
    from PyQt5.QtCore import *
elif machinery.USE_PYQT6:
    from PyQt6.QtCore import *
```

Stubs are available for both PyQt5 and PyQt6, so we'd like to run mypy twice, once with `--always-true=USE_PYQT5 --always-false=USE_PYQT6`, and once with the opposite.

There are other wrappers over PyQt5/PyQt6 which share the same approach, and usually ship a command-line tool on top of mypy to generate flags to pass to it: [qts](https://github.com/python-qt-tools/qts/blob/main/src/qts/_cli.py) (cc @altendky), [qtpy](https://github.com/spyder-ide/qtpy/blob/master/qtpy/cli.py).

However, it'd be nice to have a default setting for that (so that mypy can be run, even from e.g. an editor integration, without needing a wrapper). Then there could be a config snippet overriding those again.

Unfortunately, in our case, even having a default in the config and overriding it on the commandline won't work properly, as:

> mypy: error: You can't make a variable always true and always false (IS_QT5, IS_QT6, USE_PYQT5, USE_PYQT6)

but I suppose that's different from what this issue describes.

**Comment by ruslaniv** [comment]
Yes, this is required for monorepos badly! 
It would be great to implement it like Ruff does, where you can keep the main config file in the root and then project specific configs in the project folders and then just do:
```shell
extend = "../mypy.ini"
```

**Comment by Avasam** [comment]
I'll add that even across multiple repositories, this can enable shared based configuration through git submodules.

**Comment by ericmburgess** [comment]
Yes, please! This capability is the only thing in the way of my practice of keeping a scratch folder in each repo, which will be totally ignored by `git` and `mypy`. It's important that I can configure this _without_ modifying anything that is checked in to source control. Currently I have simply duplicated the (git tracked) `.mypy.ini` with a `mypy.ini` file, and added my personal ignores. But this is brittle since I can forget to update my replacement config.

**Comment by hasier** [comment]
Hi all! I've taken a stab at building this "the ruff way" [as mentioned above](https://github.com/python/mypy/issues/9620#issuecomment-2044571184), I hope this helps towards resolving this issue and/or re-opens the discussion on how to implement this 😊  https://github.com/python/mypy/pull/19135

**Comment by hasier** [comment]
> Hi all! I've taken a stab at building this "the ruff way" [as mentioned above](https://github.com/python/mypy/issues/9620#issuecomment-2044571184), I hope this helps towards resolving this issue and/or re-opens the discussion on how to implement this 😊 [#19135](https://github.com/python/mypy/pull/19135)

Anybody with review/write permissions in the repo watching this issue available to look at the PR? There was a tiny conflict I have now fixed, so the PR should be good for review again 🙏 (I've seen the `mypy_primer` comment but I don't think it's related to the changes, there are [other merged PRs with the same comment](https://github.com/python/mypy/pull/19050))

**Comment by wpp-vlad-ardelean** [comment]
Hi, our team is also following this with great interest. We also have a monorepo and the ability to merge a default config with more specific, per-project configs is very important for us.
I understand that the maintainers probably have a lot of work on their hands at the moment, but this has already been brought up 5 years ago and is one of the first results on google when searching for "mypy config merge". Seeing that there is already an open PR (thank you for your contribution @hasier ), likewise waiting for reviews for quite a while, I see no reason as to why this shouldn't be a higher priority.

We're looking forward to seeing this become a feature!
